src/rust/engine/graph/src/lib.rs

```diff
@@ -52,22 +52,71 @@ impl RunToken {
   }
 }
 
+///
+/// A token associated with a Node that is incremented whenever its output value has (or might
+/// have) changed. When a dependent consumes a dependency at a particular generation, that
+/// generation is recorded on the consuming edge, and can later used to determine whether the
+/// inputs to a node have changed.
+///
+/// Unlike the RunToken (which is incremented whenever a node re-runs), the Generation is only
+/// incremented when the output of a node has changed.
+///
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct Generation(u32);
+
+impl Generation {
+  fn initial() -> Generation {
+    Generation(0)
+  }
+
+  fn next(&self) -> Generation {
+    Generation(self.0 + 1)
+  }
+}
+
 enum EntryState<N: Node> {
-  NotStarted(RunToken),
+  // A node that has either been explicitly cleared, or has not yet started Running for the first
+  // time. In this state there is no need for a dirty bit because the generation is either in its
+  // initial state, or has been explicitly incremented when the node was cleared.
+  //
+  // The previous_result value is _not_ a valid value for this Entry: rather, it is preserved in
+  // order to compute the generation value for this Node by comparing it to the new result the next
+  // time the Node runs.
+  NotStarted {
+    run_token: RunToken,
+    generation: Generation,
+    previous_result: Option<Result<N::Item, N::Error>>,
```

Comment:
This could be a shared-across-instances `Arc<Mutex<Vec<Result<N::Item, N::Error>>` to get more than `n-1` re-use of generations, but unclear what usage patterns are actually going to be to know which is going to be more efficient. (If people flip up and back a lot, a `Vec` saves a lot of computation; if they tend to make linear changes, the `Mutex` and linear scan may be more costly)

(An `Arc<Mutex<HashMap<Vec<Generation>, Result<N::Item, N::Error>>>>` would come to mind, except for how we learnt about the minimum sizes of `HashMap`s)

Reply:
As discussed in Slack, I'd like to hold off on storing multiple values in this first edition. We can easily add it later if we're not seeing any unexpected memory pressure due to just holding one.